vhost-device-input/src/main.rs

```diff
@@ -0,0 +1,226 @@
+//
+// Copyright 2023 Linaro Ltd. All Rights Reserved.
+// Leo Yan <leo.yan@linaro.org>
+//
+// SPDX-License-Identifier: Apache-2.0 or BSD-3-Clause
+
+mod input;
+mod vhu_input;
+
+use log::{error, info, warn};
+use std::os::fd::AsRawFd;
+use std::process::exit;
+use std::sync::{Arc, RwLock};
+use std::thread::{self, JoinHandle};
+
+use clap::Parser;
+use thiserror::Error as ThisError;
+use vhost::{vhost_user, vhost_user::Listener};
+use vhost_user_backend::VhostUserDaemon;
+use vm_memory::{GuestMemoryAtomic, GuestMemoryMmap};
+
+use vhu_input::VuInputBackend;
+use vmm_sys_util::epoll::EventSet;
+
+type Result<T> = std::result::Result<T, Error>;
+
+#[derive(Debug, ThisError)]
+/// Errors related to vhost-device-input daemon.
+pub(crate) enum Error {
+    #[error("Event device file doesn't exists or can't be accessed")]
+    AccessEventDeviceFile,
+    #[error("Threads can't be joined")]
+    FailedJoiningThreads,
+    #[error("Could not create backend: {0}")]
+    CouldNotCreateBackend(std::io::Error),
+    #[error("Could not create daemon: {0}")]
+    CouldNotCreateDaemon(vhost_user_backend::Error),
+    #[error("Could not register input event into vring epoll")]
+    CouldNotRegisterInputEvent,
+    #[error("Fatal error: {0}")]
+    ServeFailed(vhost_user_backend::Error),
+}
+
+#[derive(Clone, Parser, Debug, PartialEq)]
+#[clap(author, version, about, long_about = None)]
+struct InputArgs {
+    // Location of vhost-user Unix domain socket.
+    #[clap(short, long)]
+    socket_path: String,
+
+    // Path for reading input events.
+    #[clap(short = 'e', long)]
+    event_list: String,
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub(crate) struct VuInputConfig {
+    pub socket_path: String,
+    pub events: Vec<String>,
+    pub count: u32,
+}
+
+impl TryFrom<InputArgs> for VuInputConfig {
+    type Error = Error;
+
+    fn try_from(args: InputArgs) -> Result<Self> {
+        let socket_path = args.socket_path.trim().to_string();
+        let event_list: Vec<&str> = args.event_list.trim().split(',').collect();
+        let mut count = 0;
+        let mut events: Vec<String> = Vec::new();
+
+        for ev_str in event_list.iter() {
+            let ev = ev_str.to_string();
+            events.push(ev.clone());
+            count += 1;
+        }
+
+        Ok(VuInputConfig {
+            socket_path,
+            events,
+            count,
+        })
```

Comment:
I will remove the field `count` since it's redundant.  

As suggested by @Ablu , I changed to use `PathBuf` for the field `events`, so it might be a bit difference from what you suggested for using `String` type.